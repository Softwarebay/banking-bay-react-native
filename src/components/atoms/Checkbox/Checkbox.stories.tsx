import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { CenterStory } from 'utils/storybook/CenterStory';

import { Checkbox } from './Checkbox';

storiesOf('Atoms/Checkbox', module)
	.addDecorator((getStory) => <CenterStory center>{getStory()}</CenterStory>)
	.add('Default', () => (
		<Checkbox testID={text('testID', 'Checkbox-test-id')} />
	));
