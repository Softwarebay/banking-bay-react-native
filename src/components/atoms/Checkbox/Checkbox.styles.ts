import { StyleSheet } from 'react-native';

import { colors } from 'theme';

export const styles = StyleSheet.create({
	checkIcon: {
		bottom: 2,
		color: colors.mainDark,
		fontSize: 22,
		right: 1,
	},
	checkbox: {
		alignItems: 'center',
		borderColor: colors.checkbox,
		borderRadius: 4,
		borderWidth: 1,
		height: 22,
		justifyContent: 'center',
		width: 22,
	},
});
