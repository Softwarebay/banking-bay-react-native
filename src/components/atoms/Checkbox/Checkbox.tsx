import { forwardRef, useImperativeHandle, useState } from 'react';
import { Pressable } from 'react-native';

import { colors } from 'theme';
import { Icon } from '../Icon/Icon';

import { styles } from './Checkbox.styles';

export interface CheckboxHandle {
	getValue: () => boolean;
	clear: () => void;
}

export interface Props {
	testID: string;
}

export const Checkbox = forwardRef<CheckboxHandle, Props>(({ testID }, ref) => {
	const [isChecked, setIsChecked] = useState(false);

	useImperativeHandle(ref, () => ({
		getValue: () => isChecked,
		clear: () => {
			setIsChecked(false);
		},
	}));

	const onPress = () => {
		setIsChecked((checked) => !checked);
	};

	return (
		<Pressable
			hitSlop={20}
			style={styles.checkbox}
			testID={testID}
			onPress={onPress}
		>
			{isChecked && <Icon color={colors.mainDark} name="check" />}
		</Pressable>
	);
});
