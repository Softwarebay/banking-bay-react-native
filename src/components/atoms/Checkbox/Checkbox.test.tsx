import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { Checkbox } from './Checkbox';

import type { Props } from './Checkbox';

describe('Checkbox', () => {
	const props: Props = { testID: 'checkbox-test-id' };

	it('should match snapshot', () => {
		const tree = testRenderer.create(<Checkbox {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should initially render unchecked', () => {
		render(<Checkbox {...props} />);

		const checkbox = screen.getByTestId('checkbox-test-id');
		const checkIcon = screen.queryByTestId('checkbox-icon-test-id');

		expect(checkbox.children).toHaveLength(1);
		expect(checkIcon).toBeNull();
	});

	it('should toggle checked state', () => {
		render(<Checkbox {...props} />);

		const checkbox = screen.getByTestId('checkbox-test-id');

		fireEvent.press(checkbox);
		const checkedIcon = screen.queryByTestId('check-icon-test-id');

		expect(checkbox.children).toHaveLength(2);
		expect(checkedIcon).toBeTruthy();

		fireEvent.press(checkbox);
		const uncheckedIcon = screen.queryByTestId('check-icon-test-id');

		expect(checkbox.children).toHaveLength(1);
		expect(uncheckedIcon).toBeNull();
	});
});
