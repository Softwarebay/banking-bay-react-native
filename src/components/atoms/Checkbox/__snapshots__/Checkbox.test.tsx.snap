// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={20}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#868698",
      "borderRadius": 4,
      "borderWidth": 1,
      "height": 22,
      "justifyContent": "center",
      "width": 22,
    }
  }
  testID="checkbox-test-id"
/>
`;
